// ng2-dnd/esm2015/src/dnd.module.js
// Copyright (C) 2016-2020 Sergey Akopkokhyants
// This project is licensed under the terms of the MIT license.
// https://github.com/akserg/ng2-dnd
var DndModule_1;
import { __decorate } from "tslib";
import { NgModule } from "@angular/core";
import { DragDropConfig } from './dnd.config';
import { DragDropService, DragDropSortableService, dragDropServiceFactory, dragDropSortableServiceFactory } from './dnd.service';
import { DraggableComponent, DraggableHandleComponent } from './draggable.component';
import { DroppableComponent } from './droppable.component';
import { SortableContainer, SortableComponent, SortableHandleComponent } from './sortable.component';
export * from './abstract.component';
export * from './dnd.config';
export * from './dnd.service';
export * from './draggable.component';
export * from './droppable.component';
export * from './sortable.component';
const ɵ0 = dragDropServiceFactory, ɵ1 = dragDropSortableServiceFactory;
export let providers = [
    DragDropConfig,
    { provide: DragDropService, useFactory: ɵ0 },
    { provide: DragDropSortableService, useFactory: ɵ1, deps: [DragDropConfig] }
];
let DndModule = DndModule_1 = class DndModule {
    static forRoot() {
        return {
            ngModule: DndModule_1,
            providers: providers
        };
    }
};
DndModule = DndModule_1 = __decorate([
    NgModule({
        declarations: [DraggableComponent, DraggableHandleComponent, DroppableComponent, SortableContainer, SortableComponent, SortableHandleComponent],
        exports: [DraggableComponent, DraggableHandleComponent, DroppableComponent, SortableContainer, SortableComponent, SortableHandleComponent],
    })
], DndModule);
export { DndModule };
export { ɵ0, ɵ1 };
