// ng2-dnd/esm2015/src/dnd.service.js
// Copyright (C) 2016-2020 Sergey Akopkokhyants
// This project is licensed under the terms of the MIT license.
// https://github.com/akserg/ng2-dnd
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { isPresent } from './dnd.utils';
export class DragDropData {
}
export function dragDropServiceFactory() {
    return new DragDropService();
}
let DragDropService = class DragDropService {
    constructor() {
        this.allowedDropZones = [];
    }
};
DragDropService = __decorate([
    Injectable()
], DragDropService);
export { DragDropService };
export function dragDropSortableServiceFactory(config) {
    return new DragDropSortableService(config);
}
let DragDropSortableService = class DragDropSortableService {
    constructor(_config) {
        this._config = _config;
    }
    get elem() {
        return this._elem;
    }
    markSortable(elem) {
        if (isPresent(this._elem)) {
            this._elem.classList.remove(this._config.onSortableDragClass);
        }
        if (isPresent(elem)) {
            this._elem = elem;
            this._elem.classList.add(this._config.onSortableDragClass);
        }
    }
};
DragDropSortableService = __decorate([
    Injectable()
], DragDropSortableService);
export { DragDropSortableService };
